### PR TITLE
Peterson impl feature finish my team filter or tasks tab

### PR DESCRIPTION
**Pull request by peterson337**
# Description
The filter was created for administrators  and owners. These users require the filter as they oversee all users listed in the Leaderboard and the Team Member Tasks tables, rendering all tools for reviewing user work useless.

## Related PRS (if any):
None
…

## Main changes explained:
The components 'Team Members' and 'Leaderboard' were modified to implement the filter.

…

## How to test:
1. check into current branch
2. do `npm install` and `...` to run this PR locally   
3. The filter is exclusively visible to administrators and owners; other users will not see the filter.
4.  Log in as an administrator or owner to verify the visibility of the filter. Log in as any other user to confirm its invisibility.
5. to test this PR you need of two or more teams.
6.  If you need to create a team, click on 'Other Links' and select 'Teams.' Then, click on 'Create New Team' and enter the team name. Next, click the button within the 'Members' column to add yourself and other users.
b
7. go to dashboard→ Tasks 
8. Click on the filter and select any team, then click on the "My Team" button.
9. After the table displays the team members, the 'My Team' button will change to the 'View All' button. When clicked, the table will display all users of the HGN software, and the 'My Team' button will reappear. Alternatively, you can select another team in the filter to display the 'My Team' button again.
10. Verify the functionality of the filter with buttons for 1 day, 2 days, 3 days, and so forth.

## Screenshots or videos of changes:
a

## Note:
Before deleting a team, ensure all members are removed. If the team is deleted first, it may still appear in the filter.

### CODE REVIEW

**Review by SanketKaware — DISMISSED** [review]
I have followed your steps as mentioned. Overall looks good.

It works as expected. 
1. When there was no team, I was unable to click on "My Team" on the dashboard.
2. After adding the team and its members, I was able to select the team and switch between "My Team" and "View All".
3. Worked for 1 day, 2 days, 3 days, and so forth.

Here is my video review for reference - 

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/17793708/d34f1074-f90a-4ea8-be56-afa613e165d5

**Review by meetpadhiar4 — DISMISSED** [review]
Hi, I tested your PR and it is working as intended. Please refer the video attached.


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/67577022/f73ccafb-fb88-4978-9def-22368ac557c3

**Review by Abi-Liu — DISMISSED** [review]
Overall the changes work as expected! I can toggle to view between my individual teams vs all users inside the leaderboard and Team member tasks section.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/59585815/a5a6b244-8cfd-4e1f-a63d-f70a59212c36

**Review by abdel-lall — COMMENTED** [review]
I've encountered an issue with the filter functionality—it's not functioning as expected. Please refer to the images below.

Before selecting a team:
Screenshot 2024-03-22 213244

After selecting a team:
Screenshot 2024-03-22 213308

Additionally, the styling of the button below doesn't seem to be as it was before.
Before:
Screenshot 2024-03-22 213357
After:
Screenshot 2024-03-22 213340

**Review by SushmithaPrathap — CHANGES_REQUESTED** [review]
Hey,
Just reviewed your PR and tested it following the steps provided.Here are my observations:

- The filter clearly indicates to create a team when none exist, which is helpful.
- Suggestion: Adding a "Create Team" button right there could streamline the process, leading users directly to the team creation page rather than navigating through the menu.
- It's great that the filter doesn't work for teams without members
- The toggle between "My Team" and "View All" feels a bit clunky
- Placing "View All" inside the dropdown is somewhat confusing. It might be more intuitive to have it separate from the team selection.
- The "My Team" blue button seems unnecessary since selecting a team from the dropdown already initiates the filter by that team. It could be more useful to repurpose this button for a "View All" functionality, simplifying the UI.
- The dashboard is going out of the view. Have to scroll horizontally

I've attached a screen recording to illustrate these points more clearly. Hope this feedback is helpful!


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/113482457/cccdbc5c-bdcc-41e6-8258-2590a9b39a08


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/113482457/dd3d9ad5-1c64-426c-a19b-83bfd80b0540

**Comment by MQChen211** [comment]
Hello, I've reviewed and tested your pull request, and everything functions as anticipated. Utilizing an Owner account, I was able to create teams, assign members to those teams, and verify that this information accurately displays on the dashboard. Additionally, when accessing the system with a Volunteer account, the filter appropriately remained hidden, aligning with the expected behavior.

Tested with a Volunteer account:
截屏2024-04-04 13 13 16

Tested with an Owner account:
截屏2024-04-04 13 06 19

截屏2024-04-04 13 08 10

截屏2024-04-04 13 09 47

截屏2024-04-04 13 12 24

截屏2024-04-04 13 12 34

**Review by MQChen211 — DISMISSED** [review]

**Review by SRamen1999 — DISMISSED** [review]
Everything works as expected. I can toggle to view between my individual teams vs all users inside the leaderboard and Team member tasks section. Also there are safe checks in place when a user does not have a team or did not select a team.


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/56932664/f9ca85b2-e9e0-4a62-9355-8a7e0f7b423a

**Comment by hiralsonip** [comment]
I've reviewed the PR, and it looks good. However, I have a couple of suggestions:

1. Instead of having multiple dropdowns, we could simplify by having just one dropdown. Within this dropdown, we can include a "View All" button for ease of navigation.
2. If a user doesn't have a team, rather than displaying the text "Please create a team," we can make the "My Team" button functional. It could allow users to create a team directly.


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/66827672/eb316d55-4f5a-4e3d-a36d-80af9b5b20bf

File: src/components/LeaderBoard/Leaderboard.jsx [hunk 1]
```diff
@@ -127,30 +211,83 @@ function LeaderBoard({
...
+                  >
+                    Please, create a team to use the filter.
+                  </DropdownItem>
+                ) : (
+                  /* eslint-disable */
+                  <>
+                    <DropdownItem onClick={() => TeamSelected(null)}>{'View All'}</DropdownItem>
+                    {teams.map(team => (
+                      <DropdownItem key={team._id} onClick={() => TeamSelected(team)}>
+                        {team.teamName}
```
Comment (anandsesha) [inline src/components/LeaderBoard/Leaderboard.jsx]: Hi, I've tested this feature and my observations are as below:
What works:
- Without adding team, the 'My Team' was not selectable as expected.
- Able to add team, add members, filter based on 'My Team' and 'View All' in dashboard.
- functionality of  filter with buttons for 1 day, 2 days and so forth works.
- Other users like CoreTeam/Volunteer can't see Team filter as expected.

**Issue**:
- When a team name with large characters is added, the same is reflected in the dashboard which disrupts the team selection view. (Refer Issue screenshots below at the end)

**Suggestion**: On line 270, recommend using `${team.teamName.substring(0, 10)}...` in the DropdownItem in the Leaderboard and under Team member task in dashboard page.

Other observations:
- After filtering based on a team in the leaderboard, when we hit refresh the list shows all users of the HGN software, even though a particular team is selected. (Please confirm if this is the expected functionality, screenshot attached)


1 Created Team
2 assigned team members
3 Team filtered in dashboard
4 viewAll expands  team
5 CoreTeam


**Issue screenshots:**
Issue - Team length 2
Issue - team length 3
Issue - Team length


Observation on leaderboard refresh:
Observ - leaderboard update with team selection

**Review by anandsesha — CHANGES_REQUESTED** [review]

**Review by anandsesha — CHANGES_REQUESTED** [review]
Hi, I've tested this feature and my observations are as below:
What works:
- Without adding team, the 'My Team' was not selectable as expected.
- Able to add team, add members, filter based on 'My Team' and 'View All' in dashboard.
- functionality of  filter with buttons for 1 day, 2 days and so forth works.
- Other users like CoreTeam/Volunteer can't see Team filter as expected.

**Issue**:
- When a team name with large characters is added, the same is reflected in the dashboard which disrupts the team selection view. (Refer Issue screenshots below at the end)

**Suggestion**: On line 270, recommend using `${team.teamName.substring(0, 10)}...` in the DropdownItem in the Leaderboard and under Team member task in dashboard page.

Other observations:
- After filtering based on a team in the leaderboard, when we hit refresh in the leaderboard, the list shows all users of the HGN software, even though a particular team is selected. (Please confirm if this is the expected functionality, screenshot attached)


1 Created Team
2 assigned team members
3 Team filtered in dashboard
4 viewAll expands  team
5 CoreTeam


**Issue screenshots:**
Issue - Team length 2
Issue - team length 3
Issue - Team length


Observation on leaderboard refresh:
Observ - leaderboard update with team selection

**Review by j-corp-25 — DISMISSED** [review]
Hi @peterson337 

I tested the feature based on the instructions and its working as expected. 
I tested with multiple buttons as well.

Below is the video;

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/95154804/d5424a4d-3bbe-4dee-9df5-ad05fafc4545

**Review by ImzIssa — DISMISSED** [review]
Hello, I've reviewed this PR and overall everything works as expected. Nice work.
PR 2078
PR 2078 (2)

For improvements consider suggestions made by @SushmithaPrathap.

**Review by p-suram — COMMENTED** [review]
Works as expected. I have tried options in "Team Member Tasks" as well. 
**Suggestions:**
1. Parallelly write unit test cases as we keep adding new features
2.  `const TeamSelected = team => {
    const teamName = `${team.teamName.substring(0, 15)}...`;
    setSelectedTeamName(team.teamName.length >= 15 ? teamName : team.teamName);
    setUsersSelectedTeam(team);
    setTextButton('My Team');
  };`
  Extract above logic into a util function so that it can be reused in other places if required. (It can be a function which takes 2 input params, title: string, length: number and return text accordingly).

Rest all looks great. Please find the recording below

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/157307140/c3a81f92-cc92-49d7-b970-a150e0dd2bd4

**Comment by peterson337** [comment]
> * When a team name with large characters is added, the same is reflected in the dashboard which disrupts the team selection view. (Refer Issue screenshots below at the end)
> 
> **Suggestion**: On line 270, recommend using `${team.teamName.substring(0, 10)}...` in the DropdownItem in the Leaderboard and under Team member task in dashboard page.
>
> 
> * After filtering based on a team in the leaderboard, when we hit refresh in the leaderboard, the list shows all users of the HGN software, even though a particular team is selected. (Please confirm if this is the expected functionality, screenshot attached)
>
>

[@anandsesha](https://github.com/anandsesha),  I fixed the bugs in your comment. Please, refer to my GIF for further reference: 
b

**Review by Parth-tech — DISMISSED** [review]
Hi Peterson,
I have checked you PR and everything looks perfect. Great job!

Checks performed:

- [x] Button visible only to admins & owners
- [x] `My Team` display the selected team
- [x] `View All` allows to view everyone
- [x] Test Leaderboard
- [x] Test for `different days` under `Tasks Tab`

Screenshot 2024-04-08 at 4 49 09 PM

Screenshot 2024-04-08 at 4 47 16 PM

Screenshot 2024-04-08 at 4 49 01 PM

Screenshot 2024-04-08 at 4 48 19 PM

Screenshot 2024-04-08 at 4 48 10 PM

Screenshot 2024-04-08 at 4 49 31 PM

I have a suggestion (for improving user experience, feel free to ignore as well):

1. Prefer hiding `My Team` button until the user has selected his team.
2. Whenever `selects a team` from the `dropdown` directly update the view automatically to display details about the team member (this is more intuitive); instead of user having to select a team from the drop down and then also click on the `My Team`

**Review by anandsesha — DISMISSED** [review]
Hi, I tested this PR again and the Issue I reported earlier is fixed now. 

- Team name with large characters is displayed correctly in Team selection view om dashboard
- Refresh leaderboard doesn't load all HGN team members and works as expected 

Great Work!

Screenshots of change attached.
Screenshot (777)
Screenshot (778)
Screenshot (779)
Screenshot (780)
Screenshot (781)

After Leaderboard refresh in a team view:
Screenshot (782)

**Review by SushmithaPrathap — DISMISSED** [review]
Hey, After reviewing the recent adjustments made to this PR, everything now works as expected. The issues previously highlighted have been effectively addressed, and the functionality aligns well with the requirements.

Here are the screenshots:

PR 2078 - 1
PR 2078 - 2

**Review by j-corp-25 — DISMISSED** [review]
Hi @peterson337 

I tested based on the instructions and the filter is working correctly. 

I have two teams under Admin and I can filter through them. I filter is also hidden from different roles. I also tested deleting a team and it doesnt show on the filter after I do this.

Screenshots and videos below:

- Admin View with Filter

AdminView


- Volunteer view with no filter view

VolunteerView

- Video following testing filter before and after deletion of teams


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/95154804/a568ad52-4fde-4cdd-a87b-4f2631a15726

**Review by p-suram — COMMENTED** [review]
Changes look good working as expected. Filter is only visible for Admin, Owner users but not for Volunteer user. I have 2 minor suggestions:
1. Lines 224-225 of Leaderboard.jsx can be extracted into separate util function as this logic is being repeated quite a number of times.
2. I see team name as `Pooja_test_team...` even though team name is `Pooja_test_team`. Please find video below:

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/157307140/14614f4f-5609-434b-84d8-dee46ffec51a

**Review by metaphor987 — COMMENTED** [review]
Hi Peterson,

I've tested the changes, and the main functionalities all work well. The team filter is only visible to admin accounts and not volunteer accounts, which works as expected.

The only issue I encountered is that when clicking on the "Tracking" button for a team member in "Team Member Tasks" on the left side (regardless of whether the member belongs to the admin's team or not), the "Leaderboard" on the right side automatically changes to show all members' info, even though we are still in "My team" mode. I'm not sure if this action is expected. Thank you!

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/91745551/83e345a2-0e55-4ca9-bad0-d8c9c6ba6978

**Review by Tiantian-C — COMMENTED** [review]
I have tested it and it works as intended.
Screenshot 2024-04-13 at 5 44 15 PM
Screenshot 2024-04-13 at 5 44 44 PM

**Review by p-suram — CHANGES_REQUESTED** [review]
Changes look good working as expected. Filter is only visible for Admin, Owner users but not for Volunteer user. I have 2 minor suggestions:
1. Lines 224-225 of Leaderboard.jsx can be extracted into separate util function as this logic is being repeated quite a number of times.
2. I see team name as `Pooja_test_team...` even though team name is `Pooja_test_team`. Please find video below:

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/157307140/14614f4f-5609-434b-84d8-dee46ffec51a

**Comment by peterson337** [comment]
> Hi Peterson,
> 
> I've tested the changes, and the main functionalities all work well. The team filter is only visible to admin accounts and not volunteer accounts, which works as expected.
> 
> The only issue I encountered is that when clicking on the "Tracking" button for a team member in "Team Member Tasks" on the left side (regardless of whether the member belongs to the admin's team or not), the "Leaderboard" on the right side automatically changes to show all members' info, even though we are still in "My team" mode. I'm not sure if this action is expected. Thank you!
> 
Hello [@metaphor987](https://github.com/[metaphor987](https://github.com/metaphor987)), I fixed the bug in your comment. Please, refer to my GIF for further reference:

**Review by metaphor987 — DISMISSED** [review]
Hi Peterson,

Thank you for the updates! The GIF you have attached is really clear. I've tested the changes on my side and everything looks great. Here's a screen recording showing that the previous bug has been fixed.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/91745551/28d4868a-249b-480b-8c57-6bae10035099


Thank you!

**Review by p-suram — DISMISSED** [review]
Hey Peterson, Thank you addressing the comments. Code changes look good and working as expected. Please find attached video that shows that previous issue has been resolved.


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/157307140/f0b3273e-d437-4252-90d8-b0c05583c844

Great Work!

**Review by Abi-Liu — DISMISSED** [review]
Hey Peterson! The changes work as expected. I'm able to toggle between teams only on an admin/owner account.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/59585815/9b88958b-64b4-4e61-8738-bf1ba6dc8051

**Review by meetpadhiar4 — DISMISSED** [review]
Hi, i tested your PR and it is working as intended.

**Review by abdel-lall — COMMENTED** [review]
Hello, @peterson337. I wanted to bring to your attention that the filter on the task board is still not displaying the tasks and their progress, which is crucial for the task board's functionality. Additionally, it seems that the style of the time-off button is still not quite right. Everything else looks good.
Screenshot 2024-05-04 135046
Screenshot 2024-05-04 135056

**Comment by peterson337** [comment]
> Hello, @peterson337. I wanted to bring to your attention that the filter on the task board is still not displaying the tasks and their progress, which is crucial for the task board's functionality. Additionally, it seems that the style of the time-off button is still not quite right. Everything else looks good. Screenshot 2024-05-04 135046 Screenshot 2024-05-04 135056

Hello [abdel-lall](https://github.com/abdel-lall), I noticed that you're testing the old version of my pull request. The filter that displays the user's tasks.  
x
If you test my pull request again, please use 'git pull' in your terminal.

**Review by cgomezhub — CHANGES_REQUESTED** [review]
Good job! It seems like you're doing very well, almost everything is alright, However, check the time recording feature, it is not showing correctly on the dashboard.

Screenshot 2024-05-06 144346
Screenshot 2024-05-06 144837
Screenshot 2024-05-06 151506
Screenshot 2024-05-06 151811

**Review by SRamen1999 — DISMISSED** [review]

**Review by ImzIssa — DISMISSED** [review]
Hello, I've tested your PR and the filter works as expected. Great Work.

Filter Invisible when logged in as Volunteer
PR2078

Filter works on Tasks and Leaderboard when logged in as Admin
PR2078 (2)

**Review by abdel-lall — DISMISSED** [review]
LGTM. Good job!
Screenshot 2024-05-11 122827

**Comment by one-community** [comment]
> Good job! It seems like you're doing very well, almost everything is alright, However, check the time recording feature, it is not showing correctly on the dashboard.
> 
> Screenshot 2024-05-06 144346 Screenshot 2024-05-06 144837 Screenshot 2024-05-06 151506 Screenshot 2024-05-06 151811

This looks correct to me, what are you seeing that is wrong?

**Review by cgomezhub — DISMISSED** [review]
I had some issues with charging but the functionality works as expected

https://www.loom.com/share/49dd2825adcb4241a273610d33996bd5

**Review by nathcarnevalli — DISMISSED** [review]
The functionality worked as expected for different type of accounts.

Screenshot 2024-05-16 104136

**Comment by nishithaShetty** [comment]
Hey Peterson, nice work. The functionality works as expected.
It's not related to the functionality but the inactive button in the delete team section, does not work as expected.

PR#2078
PR#2078_1

**Review by Sandhya1236 — DISMISSED** [review]
I checked out the current branch, ran npm install and the necessary commands to run the PR locally. I verified that the filter is exclusively visible to administrators and owners by logging in as an administrator and owner, confirming the filter's visibility. I also logged in as a regular user to ensure the filter was not visible. To test with two or more teams, I created new teams by navigating to 'Other Links' → 'Teams', then clicked 'Create New Team', entered the team name, and added members including myself. All steps were followed successfully, confirming the filter visibility works as intended based on user roles.

image

**Comment by one-community** [comment]
Thank you all, moving to final review.

**Comment by ZhangPeizhou** [comment]
https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/95693786/4370426b-e407-4c6f-97ee-41a76fac649e

Work as expected, well done!

**Review by vishavk1992 — APPROVED** [review]
I have tested this PR by following the instructions and it works as intended . 
After adding the team and its members, I was able to select the team and switch between "My Team" and "View All".
Worked for 1 day, 2 days, 3 days, and so forth.


https://github.com/user-attachments/assets/0e405c5c-e47a-4f6e-8037-8ca84d6787b0

**Comment by one-community** [comment]
Thank you all, merging!

**Pull request merged**